cas-server-core-web/src/main/java/org/jasig/cas/web/flow/AbstractNonInteractiveCredentialsAction.java

```diff
@@ -68,17 +78,21 @@ protected final Event doExecute(final RequestContext context) {
         final String ticketGrantingTicketId = WebUtils.getTicketGrantingTicketId(context);
         final Service service = WebUtils.getService(context);
 
-        if (isRenewPresent(context)
-            && ticketGrantingTicketId != null
-            && service != null) {
+        if (isRenewPresent(context) && ticketGrantingTicketId != null && service != null) {
 
             try {
+                final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
```

Comment:
Sorry to be so "painful". And you're right: SWF is not related here.

Trying to pushing my ideas once again:

1) I think credentials cleaning should go into the `AuthenticationManager`: it's a few lines of source code and I don't see where they could be used otherwise. So I would remove all the stuffs around: `AuthenticationTransactionManager` and `AuthenticationTransactionFactory`

2) We need to have some authentication context, add authentications to it and then generate a final authentication (using the `PrincipalElectionStrategy`).

3) Do the `AuthenticationContext` really need to carry the `Service`?

Let's try to rewrite this part. I assume the `PrincipalElectionStragegy` is injected in the class as well as `AuthenticationManager` (maybe this is where we need your repository to gather both):

```
Authentication authentication = authenticationManager.authenticate(credential);
AuthenticationContext context = new DefaultAuthenticationContext();
context.collect(authentication);
Authentication result = context.buildFinalAuthentication(principalElectionStrategy);
```

Does it make a little more sense?